components/fxa-client/src/oauth.rs

```diff
@@ -167,25 +166,49 @@ impl FirefoxAccount {
     /// Fetch an OAuth code for a particular client using a session token from the account state.
     /// This method doesn't support OAuth public clients at this time.
     ///
-    /// * `client_id` - OAuth client id.
-    /// * `scopes` - Space-separated list of requested scopes.
-    /// * `state` - OAuth state.
-    /// * `access_type` - Type of OAuth access, can be "offline" and "online.
+    /// * `auth_params` Authentication parameters
     pub fn authorize_code_using_session_token(
         &self,
-        client_id: &str,
-        scope: &str,
-        state: &str,
-        access_type: &str,
+        auth_params: AuthorizationParameters,
     ) -> Result<String> {
         let session_token = self.get_session_token()?;
+        let (code_challenge, code_challenge_method, keys_jwe) = if let Some(key_params) =
+            auth_params.key_params
+        {
+            let keys_jwk = base64::decode_config(key_params.keys_jwk, base64::URL_SAFE_NO_PAD)?;
+            let keys_jwk = String::from_utf8(keys_jwk)?;
+            let requested_scopes: HashSet<&str> = auth_params.scope.split_whitespace().collect();
+            let data: HashMap<String, ScopedKey> = self
+                .state
+                .scoped_keys
+                .iter()
+                .filter(|(scope, _)| requested_scopes.contains(&(*scope).as_ref()))
```

Comment:
Oh! This made we realize something which I don't think we've thought about explicitly yet.

There are two important questions to consider here when handing out keys to another client:
1) Do we have all the keys that the client requested?
2) Is that client actually *allowed* to request those keys?

The corresponding function in Desktop is [here](https://searchfox.org/mozilla-central/rev/31d8600b73dc85b4cdbabf45ac3f1a9c11700d8e/services/fxaccounts/FxAccountsKeys.jsm#224) and it has two parts:

1) The call to `getScopedKeyData`, which asks the FxA server what keys the given `clientId` is actually allowed to request.
2) The call to `getKeyForScope` which retrieves the key from our local store so it can be returned to the client.

I think we can defer this to a follow-up bug, but could you please add a "TODO" comment here to note that we need this extra check, and file a follow-up bug for next sprint?

(The reason I'm asking for you to file the bug, is so I can check whether what I said above made sufficient sense to enable you to file it, so please ask clarifying questions if necessary 😁)

Reply:
Added the TODO in the code, and before I file the issue, lemme summarize (for my own sanity):
- Right now (in this pr) I simply run around the state and get any keys that fit the requested scopes
- However, the client calling us, may not be allowed to request those scopes, so, we have to check with the server first, and it would tell us which scopes the client is allowed to request.
- Then, for each of those scopes, we look internally and we basically do what we have here (i.e the whole keys_jwe encryption business)

One last crazy question, if we (as in the fxa-client) do not have the requested (and validated) scoped keys in our state, I'm guessing we do the whole generation using HKDF on our `kB` and the scope (or do we just error out)?

Reply:
> However, the client calling us

Re-phrasing this just in an attempt to be super clear: "the client from which we received a pairing request", which in practice is going to be a phone somewhere that scanned a QR code we were showing, and about which we know absolutely nothing other than what information is provided in the AuthorizationParameters.

> may not be allowed to request those scopes, so, we have to check with the server first, and it would tell us
> which scopes the client is allowed to request.

Correct. We'll provide the requested scopes to the server, and the server will respond with a filtered subset of those scopes (possibly empty!) along with the data required to generate the derived keys:

  * https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/docs/oauth/api.md#response-11

We already have the derived keys so we don't need to "generate" them as such, but we can use this as permission check on whether that client is supposed to be requesting those keys. (Of course, the client could *lie* about who it is, e.g. a malicious app could tell us that it's Fenix by providing the Fenix `client_id`. But this check serves as a defense-in-depth measure so that an attacker who e.g. compromised the Firefox Send oauth flow, couldn't use it to request the sync keys).

> Then, for each of those scopes, we look internally and we basically do what we have here 

Right. We'd basically filter the current `requested_scopes` into an `allowed_scopes` and then do the same thing we have here already.

> if we (as in the fxa-client) do not have the requested (and validated) scoped keys in our state,
> I'm guessing we do the whole generation using HKDF on our kB and the scope (or do we just error out)?

If we have `kB`, we could do this, and maybe it would be useful for tests. But in practice we don't persist `kB` so we won't be able to do it in real applications. The right thing here is probably to return an error, since we cannot fulfill the authorization request as given.